default.nix

```diff
@@ -0,0 +1,50 @@
+{ # The git revision here corresponds to the nixpkgs-unstable channel, which at
+  # the time of this writing has GHC 8.6.5 as the default compiler (matching the
+  # one used by stack.yaml). Use https://howoldis.herokuapp.com/ to determine
+  # the current rev.
+  pkgs ? import (builtins.fetchTarball "https://github.com/nixos/nixpkgs/archive/002b853782e.tar.gz") {}
+  # Which GHC compiler to use.
+  # To determine the list of compilers available run:
+  #   nix-env -f "<nixpkgs>" -qaP -A haskell.compiler
+, compiler ? "default"
+}:
+let
+  haskellPackages =
+    if compiler == "default"
+      then pkgs.haskellPackages
+      else pkgs.haskell.packages.${compiler};
+  fetchGitHubArchive = owner: repo: rev:
+    builtins.fetchTarball "https://github.com/${owner}/${repo}/archive/${rev}.tar.gz";
+
+  # Summoner project derivation.
+  projectDrv = haskellPackages.extend (pkgs.haskell.lib.packageSourceOverrides {
+    summoner = ./summoner-cli;
+    summoner-tui = ./summoner-tui;
+
+    relude = fetchGitHubArchive "kowainik" "relude"
+      "55968311244690f5cc8b4484a37a63d988ea2ec4";
+    tomland = fetchGitHubArchive "kowainik" "tomland"
+      "8ef78e5fcfd7055b1db6402713d27a0aa42a82f8";
+    shellmet = fetchGitHubArchive "kowainik" "shellmet"
+      "36149eb0eb2b81916a93cdb92f3cb949d2eb9d23";
+    optparse-applicative = fetchGitHubArchive "pcapriotti" "optparse-applicative"
+      "5478fc16cbd3384c19e17348a17991896c724a3c";
+    parser-combinators = fetchGitHubArchive "mrkkrp" "parser-combinators"
+      "4262c00ef70cc30bfc56db0a10c37d88ad88fe1a";
```

Comment:
I wonder if we would use `LTS-14.1` where all of these packages are present, we won't need these lines anymore? (Sorry, I'm not using Nix, so I don't know how it works...)

Reply:
I also wonder why all these source dependencies are required? If it's possible to update dependencies in `summoner` and simplify `nix` config, I'm all for it. I also would like to know how to determine whether I need such source dependencies or not :slightly_smiling_face:

Reply:
@vrom911 Here's what I gathered from chatting with nix folks on fpslack:

> It's a bit of a tangled process. One part of it is all-cabal-hashes, which is occasionally updated from here: https://github.com/commercialhaskell/all-cabal-hashes/releases
Part of it is hackage2nix: https://github.com/NixOS/cabal2nix/tree/master/hackage2nix (edited) 
https://github.com/NixOS/nixpkgs/commit/5ad0677aec2455e664992d93d063d5cb34a32e2f
And then part of it is this: https://github.com/NixOS/nixpkgs/blob/5ad0677aec2455e664992d93d063d5cb34a32e2f/pkgs/development/haskell-modules/configuration-hackage2nix.yaml
Which I suppose is generated from an LTS, probably

(That doesn't look like a satisfactory explanation unfortunately, but it is what it is ...)

The only way to modify the package set is to either fork nixpkgs ourselves, or wait for upstream to update it. In either case, nixpkgs will be pinned to a specific git revision (for reproduciblity).

@chshersh 
> I also wonder why all these source dependencies are required? 

They are required (except relude and shellmet, which I've just removed), because the version in pinned nixpkgs is either older (than what the cabal file asks for) or marked as broken.

>  I also would like to know how to determine whether I need such source dependencies or not 

Nix will complain about missing or broken packages. 

For example, running `nix-build` (without specifying these overrides) produced this:

![image](https://user-images.githubusercontent.com/3998/63272822-c6414900-c26a-11e9-906e-33e7d6be6d96.png)

`tomland` is marked as broken; so I went and did an override of it in default.nix.

And elsewhere it produced this:

![image](https://user-images.githubusercontent.com/3998/63272963-0acce480-c26b-11e9-94d5-9aaf8e1d5526.png)

The version of `parser-combinators` in the pinned nixpkgs is too old; so I did an override of it in default.nix by using a version that matches the version bounds asked by the cabal file.

IIUC this is similar to the use of `extra-deps` in Stack.

Reply:
Interesting. Thanks for writing this up!
I guess we should update the pinned version of `nixpkgs` with every new GHC releases or it better to update it more frequent?

Reply:
> I guess we should update the pinned version of nixpkgs with every new GHC releases or it better to update it more frequent?

I think updating for every new GHC release is good. A newer nixpkgs may also have newer versions of Haskell packages.

As long as `compiler == "default"` - the Haskell packages will be downloaded from nix cache instead of needing to build locally.